pkg/controller/machineset/machineset_controller.go

```diff
@@ -0,0 +1,191 @@
+	// Create a new controller
+	c, err := controller.New("machineset-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to MachineSet
+	err = c.Watch(&source.Kind{Type: &machinev1beta1.MachineSet{}},
+		&handler.EnqueueRequestForObject{}, predicate.ResourceVersionChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	mapper := msmapper{client: mgr.GetClient()}
+	err = c.Watch(&source.Kind{Type: &bmh.BareMetalHost{}},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper}, predicate.ResourceVersionChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileMachineSet{}
+
+// ReconcileMachineSet reconciles a MachineSet object
+type ReconcileMachineSet struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a MachineSet object and makes changes based on the state read
+// and what is in the MachineSet.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=cluster.k8s.io,resources=machinesets,verbs=get;list;watch;update;patch
+func (r *ReconcileMachineSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := log.WithValues("MachineSet", request.NamespacedName.String())
+	ctx := context.TODO()
+	// Fetch the MachineSet instance
+	instance := &machinev1beta1.MachineSet{}
+	err := r.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Don't take action if the annotation is not present
+	annotations := instance.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		return reconcile.Result{}, nil
+	}
+	_, present := annotations[AutoScaleAnnotation]
+	if !present {
+		return reconcile.Result{}, nil
+	}
+
+	// Make sure the MachineSet has a non-empty selector.
+	msselector, err := metav1.LabelSelectorAsSelector(&instance.Spec.Selector)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if msselector.Empty() {
+		// The cluster-api machinesetcontroller expects every MachineSet to have
+		// its Selector set.
+		log.Info("MachineSet has empty selector, which is unexpected. Will not attempt scaling.")
+		return reconcile.Result{}, nil
+	}
+
+	hostselector, err := actuator.SelectorFromProviderSpec(&instance.Spec.Template.Spec.ProviderSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	hosts := &bmh.BareMetalHostList{}
+	opts := &client.ListOptions{
+		Namespace: instance.Namespace,
+	}
+
+	err = r.List(ctx, hosts, client.UseListOptions(opts))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var count int32
+	for _, host := range hosts.Items {
```

Comment:
If this loop was moved to a function that returned a slice with all of the hosts that should be counted for a MachineSet, then the unit tests could more precisely test the cases like hosts being used by a set but no longer matching the host selector or matching but with a consumerRef pointing to something else.

Reply:
I avoided that because it adds complexity in two ways, but you're right that it makes testability a lot easier, and I do think it's worth it in this case. I broke out the matching logic into a separate function.

You end up passing a lot of stuff in. The function definition I ended up with is:

```
func (r *ReconcileMachineSet) hostMatches(ctx context.Context, hostselector labels.Selector,
	msselector labels.Selector, host *bmh.BareMetalHost) (bool, error)
```

You also end up having to do another layer of error abstraction, which looks like this:

```
var count int32
for i := range hosts.Items {
	matches, err := r.hostMatches(ctx, hostselector, msselector, &hosts.Items[i])
	switch {
	case err == errConsumerNotFound:
		return reconcile.Result{}, nil
	case err != nil:
		return reconcile.Result{}, err
	case matches == true:
		count++
	}
}
```